blockdevice/probe/probe.go

```diff
@@ -34,106 +26,96 @@ type ProbedBlockDevice struct {
-		return nil, err
-	}
-
-	for _, info := range infos {
-		devpath := "/dev/" + info.Name()
+// SelectOption is a callback matcher for All block devices probes.
+type SelectOption func(device *ProbedBlockDevice) (bool, error)
 
-		var probed []*ProbedBlockDevice
-
-		probed, err = probeFilesystem(devpath)
+// WithPartitionsLabeled search for a block device which has partitions with some specific labels.
+func WithPartitionsLabeled(label string) SelectOption {
+	return func(device *ProbedBlockDevice) (bool, error) {
+		pt, err := device.PartitionTable()
 		if err != nil {
-			return nil, err
+			return false, err
 		}
 
-		all = append(all, probed...)
+		return pt.Partitions().FindByName(label) != nil, nil
 	}
-
-	return all, nil
 }
 
-// FileSystem probes the provided path's file system.
-func FileSystem(path string) (sb filesystem.SuperBlocker, err error) {
-	var f *os.File
-	// Sleep for up to 5s to wait for kernel to create the necessary device files.
-	// If we dont sleep this becomes racy in that the device file does not exist
-	// and it will fail to open.
-	err = retry.Constant(5*time.Second, retry.WithUnits((50 * time.Millisecond))).Retry(func() error {
-		if f, err = os.OpenFile(path, os.O_RDONLY|unix.O_CLOEXEC, os.ModeDevice); err != nil {
-			if os.IsNotExist(err) {
-				return retry.ExpectedError(err)
-			}
-
-			return retry.UnexpectedError(err)
-		}
+// All probes a block device's file system for the given label.
+func All(options ...SelectOption) (all []*ProbedBlockDevice, err error) {
+	var infos []os.FileInfo
 
-		return nil
-	})
-	if err != nil {
+	if infos, err = ioutil.ReadDir("/sys/block"); err != nil {
 		return nil, err
 	}
 
-	//nolint: errcheck
-	defer f.Close()
+	for _, info := range infos {
+		devpath := "/dev/" + info.Name()
 
-	superblocks := []filesystem.SuperBlocker{
-		&iso9660.SuperBlock{},
-		&vfat.SuperBlock{},
-		&xfs.SuperBlock{},
-	}
+		probed := probePartitions(devpath)
 
-	for _, sb := range superblocks {
-		if _, err = f.Seek(sb.Offset(), io.SeekStart); err != nil {
-			return nil, err
-		}
+		for _, part := range probed {
+			add := true
+			for _, matches := range options {
+				add, err = matches(part)
+				if err != nil {
+					return nil, err
+				}
 
-		err = binary.Read(f, binary.BigEndian, sb)
-		if err != nil {
-			return nil, err
-		}
+				if !add {
+					break
+				}
+			}
 
-		if sb.Is() {
-			return sb, nil
+			if add {
+				all = append(all, probed...)
+			}
 		}
 	}
 
-	return nil, nil
+	return all, nil
 }
 
 // GetDevWithFileSystemLabel probes all known block device's file systems for
 // the given label.
 func GetDevWithFileSystemLabel(value string) (probe *ProbedBlockDevice, err error) {
 	var probed []*ProbedBlockDevice
 
-	if probed, err = All(); err != nil {
+	if probed, err = All(WithPartitionsLabeled(value)); err != nil {
 		return nil, err
 	}
 
-	return filterByLabel(probed, value)
+	if len(probed) == 0 {
+		return nil, os.ErrNotExist
+	}
+
+	return probed[0], nil
```

Comment:
should we return an error if several partitions are found?

Reply:
yeah, let's do it